content/article/why-i-hate-json-and-you-can-too-.md

```diff
@@ -0,0 +1,177 @@
+entails a significant level of complexity for an encoder: a JSON encoder
+cannot encode a string without first inspecting it to look for characters
+that need to be escaped. A string of _n_ bytes can thus be up to _2n_ bytes in
+length, and the only way to know for certain is to inspect each individual
+byte in the string.
+
+It’s far more efficient—particularly for documents that are primarily
+strings—to use a serialization format that encodes string length directly
+and thus obviates any need to escape strings. Both encoding and decoding then
+become a simple copy from one buffer to another.
+
+# 6. Iffy Standardization
+
+I’ve been linking in this article to [json.org](https://json.org) rather than
+a formal
+reference for the format because there **is**, in fact, no unified JSON
+standard. Both the IETF and ECMA maintain JSON standards—the IETF has
+issued no fewer than **three** recensions of theirs!
+
+Even putting that aside, JSON has some peculiar encoding nits. Contrary
+to widespread belief, JSON is _not_ a subset of JavaScript/ECMAScript
+because JSON, unlike JavaScript, permits Unicode line and paragraph
+separators (U+2028 and U+2029, respectively).
+
+JSON is also—again, contrary to widespread misconception—_not_ a subset of
+[YAML](https://yaml.org/), another widely-used serialization format. YAML,
+for example, does not allow the `\/` sequence in encoded strings; this
+sequence is common (though not required) in JSON encoders to avoid HTML
+encoding problems when JSON is output directly into the body of an HTML
+`<script>` tag.
+
+So, JSON is its own little beast.
+
+What to Use Instead?
+=======================
+
+I find [TOML](https://github.com/toml-lang/toml) a compelling format for
+configuration files. It fixes JSON’s UTF-16 problem by allowing both
+`\uXXXX` and `\UXXXXXXXX` formats, and it also allows for comments.
+Projects as diverse as Rust’s [Cargo](https://doc.rust-lang.org/cargo/)
+package manager and [Hugo](https://gohugo.io/)—which powers this site!—are
+some of TOML’s most prominent adopters.
+
+For applications I recommend either the IETF’s [CBOR](https://cbor.io/)
+or Booking.com’s [Sereal](https://github.com/Sereal/Sereal). (Disclaimer:
+I wrote 2 of CPAN’s 3 CBOR libraries.) Both of these
+offer improved efficiency compared to JSON and,
+critically, can store binary strings. CBOR is probably better for applications
+that need to communicate across multiple languages, whereas Sereal boasts
+especially good Perl support, including storage of regular
+expressions—with modifiers!
+
+(Another format, [MessagePack](https://msgpack.org/), is a forerunner to
+CBOR in many respects. MessagePack remains in wide use but
+has only [one CPAN implementation](https://metacpan.org/pod/Data::MessagePack),
+which appears to be unmaintained.)
+
+This all having been said: if you are confident that your application
+will _never_ need to serialize anything that isn’t UTF-8, and if human
```

Comment:
A bit of a weird statement because data isn't anything. UTF-8 is just an encoding.

Reply:
UTF-8 encodes characters, though, not bytes. If the application wants to send a string of bytes, JSON can’t oblige unless those bytes happen to be exclusively UTF-8 encoded characters.

Reply:
Sure, but if your complaint is that a text format doesn't handle binary data, you have a long list of formats you should also hate. Singling out one format for that complaint is a bit misleading.

Reply:
Huh, this seems to work:

	use v5.10;

	use Socket qw(inet_aton inet_ntoa);
	use Mojo::JSON qw(encode_json decode_json);

	my $ip = '255.7.0.2';
	my $n = inet_aton( $ip );

	$perl = { ip => $n };

	my $json = encode_json( $perl );

	say $json;

	my $decoded = decode_json( $json );

	say "IP is ", inet_ntoa( $decoded->{ip} );

I know it's a simple example, which is why I'm pushing you to show the breaking case.

Reply:
To be honest, I’m a bit surprised that that works. I need to research this a bit more, but what appears to be happening is that Mojo::JSON treats every octet of a Perl string that lacks the utf8 flag as if it were a Unicode code point. So we end up with a Unicode string of all U+0000 - U+00FF. That may work in a closed ecosystem, but then we gets stuff like:
```
> perl5.30 -Mstrict -Mwarnings -MJSON -e'syswrite \*STDOUT, "[" . JSON::encode_json("\xff\xc4\x80\0") . "]"' | jq
[
  "ÿÄ\u0000"
]
```
The `\xff` octet gets transformed into U+00FF, but `\xc4\x80`, which is U+0100 (Ā) in UTF-8, comes out mangled because it’s converted to U+00C4 and U+0080.